### PR TITLE
Add `fromdict` function and `attr.ib(type)` property

### PR DESCRIPTION
**Pull request by ChrisNeveu**
This PR adds a new `type` property to `Attribute`s. I've also included in this PR a motivating example for this addition in the form of a `fromdict` function based on the `type` field. The `type` property could also be useful for possible future integration into mypy. I've never contributed to attrs before so I apologize for the large size of this changeset.

I can split the `type` addition and `fromdict` addition into two PRs if you'd like, I mostly included `fromdict` to show how the addition of the `type` property could be used.

Please let me know if there are any changes you'd like to see to bring this more in line with your overall philosophy for attrs.

### CODE REVIEW

File: src/attr/_make.py [hunk 1]
```diff
@@ -248,8 +251,13 @@ def attributes(maybe_cls=None, these=None, repr_ns=None,
 
         ..  _slots: https://docs.python.org/3.5/reference/datamodel.html#slots
 
+    :param bool validate_types: Check that each value passed to the constructor
```
Comment (ChrisNeveu) [inline src/attr/_make.py]: This reduces redundancy in having to specify both a type and the `instance_of` validator for that type.

**Comment by ChrisNeveu** [comment]
Apologies for the test failures, there is a slight difference in `typing` between 3.5.2 and 3.5.1 (which I was using locally).

**Comment by hynek** [comment]
Hi Chris thank you for the work you’ve put into this!  Have you seen the rainbow bikeshed on this topic #12?  We haven’t decided on the ideal way to go this way so it’s too early to put time and sweat into implementations. :(

**Comment by ChrisNeveu** [comment]
I've read through that discussion. I think the added `type` field is a great way to resolve some of the concerns in that thread regarding `instance_of`, but even outside of `fromdict` it would be nice to have the field for eventual mypy integration (I'm working on a plugin).

If you're willing to merge just the addition of a `type` field, let me know and I'll remove `fromdict`.

**Comment by hynek** [comment]
I'm honestly sorry for the effort you put into this but for now, http://cattrs.readthedocs.io/ is the playground for this kind of stuff.

**Pull request closed without merging**